Add new annotation file format ajava files

jwaataja opened the pull request:
New pull request since CI stopped on the old one.

Adds support for the ajava file format, as described in [this](https://docs.google.com/document/d/1ZMIIEVr6C0_SuXw1BpY7yFCw_cChVYkCc0oJh12DhCA/edit?usp=sharing) document.

In short, this file format extends the capabilities of stub files by allowing annotations in anonymous classes and local classes to be read, but the ASTs of the ajava file and any source file they apply to must match up to annotations. In the future, this format could be extended to read annotations in more locations stub files wouldn't be able to support.

There's a new implementation of `WholeProgramInference` called `WholeProgramInferenceJavaParser` that behaves identically to `WholeProgramInferenceScenes` but uses ajava files and supports anonymous inner classes.

The main component is `JointJavacJavaParserVisitor`. This traverses a javac tree and JavaParser AST simultaneously (that represent the same Java file). This allows parsing ajava files with JavaParser and then relating its nodes to the javac tree of the same file.

mernst commented:
@jwaataja Could you please pull master into your branch?  A [recent commit](https://github.com/typetools/checker-framework/commit/b6acf04747f2e0309cf9a8432793a2fbf83a4b5c) moved the WPI tests from the `checker` project to the `framework` project.

mernst commented on framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java:
```diff
@@ -623,34 +705,46 @@ private void processPackage(PackageDeclaration packDecl) {
      *     otherwise null
+     * @param classTree the tree corresponding to typeDecl if processing an ajava file, null
+     *     otherwise
+     * @return a list of types variables for {@code typeDecl}. These should be removed after
```
Removed from where?

Also, many clients ignore the return value.  Is that OK?  Is ignoring the return value OK exactly when the last argument is null?  Or maybe when isParsingStubFile is false:?

Should a similar note be added to `processCallableDeclaration`?

mernst reviewed (changes requested):
Could you please address the "TODO JWAATAJA" comments in WholeProgramInferenceJavaParser.java?

There are new prepare*ForWriting methods in WholeProgramInferenceScenes.
Could you please implement similar ones in WholeProgramInferenceJavaParser?

There is also an override of prepareMethodForWriting in FormatterAnnotatedTypeFactory.java.
That file will need to define a nested class FormatterWholeProgramInferenceJavaParser.

mernst commented on framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java:
```diff
@@ -682,6 +763,37 @@ public void setRoot(@Nullable CompilationUnitTree root) {
+
+        if (root != null && checker.hasOption("ajava")) {
+            // TODO: This relies on the toString implementations of MemberSelectTree and
+            // IdentifierTree. This could be turned into its own method that does this correctly.
```
What is "this" in "does this correctly"?  Can you add a brief description of what this code does?

mernst commented on framework/src/main/java/org/checkerframework/framework/ajava/AjavaUtils.java:
```diff
@@ -0,0 +1,50 @@
+     * @param element the type element to look at
+     * @return path to the source file containing {@code element}
+     */
+    public static String getSourceFilePath(TypeElement element) {
```
This method does not seem specific to JavaParser.  I would put it in ElementUtils.

mernst commented on framework/src/main/java/org/checkerframework/framework/ajava/AjavaUtils.java:
```diff
@@ -0,0 +1,50 @@
+     * @param name name of a top level type declaration in {@code root}
+     * @return a top level type declaration in {@code root} named {@code name}
+     */
+    public static TypeDeclaration<?> getTypeDeclarationByName(CompilationUnit root, String name) {
```
For any method or class that is an operation just over JavaParser objects, I suggest putting them in a new `JavaParserUtil` class.
That would include this methad and also `ClearAnnotationVisitor` and `StringLiteralConcatenateVisitor`.
That would not include `AnnotationConversion` and `AnnotationTransferVisitor`, because they walk two data structures in tandem.

mernst commented on framework/src/main/java/org/checkerframework/framework/ajava/ExpectedTreesVisitor.java:
```diff
@@ -0,0 +1,421 @@
+ * traverses a javac tree and JavaParser AST simultaneously, so the trees this class stores can be
+ * used to test if the entirety of the javac tree was visited.
+ *
+ * <p>Adds these trees to a set during traversal. Many trees shouldn't be matched with a JavaParser
```
You don't need to list them here, but can you say how to find, in the source code below, all the trees that shouldn't be matched with a JavaParser node?  Is it just a search for the string "javac"?

jwaataja replied:
I don't quite understand what you mean. The class as a whole contains the instructions for how to find trees that shouldn't be matched with a JavaParser node. These exceptions occur at each call to `trees.remove` and when some visit method is overriden and a child is not visited. In that case the entirety of the child's subtree tree shouldn't be matched to JavaParser nodes. As far as I'm aware, all such exceptions are accompanied by a comment.

mernst commented on framework/src/main/java/org/checkerframework/framework/ajava/StringLiteralConcatenateVisitor.java:
```diff
@@ -0,0 +1,38 @@
+    public void visit(BinaryExpr node, Void p) {
+        super.visit(node, p);
+        if (node.getOperator() == BinaryExpr.Operator.PLUS
+                && node.getRight().isStringLiteralExpr()) {
```
By experimentation, I verified that javac returns the literal string "ab" for this code:
```java
    return ("a") + ("b");
```

Please make this implementation handle parenthesization, and add a test.

mernst replied:
I observed that optimization in the generated bytecode.  But @jwaataja does not observe that in the AST.  So nothing needs to be done here (except maybe add a comment explaining why this code does not look within parentheses).

mernst commented on framework/src/main/java/org/checkerframework/framework/ajava/ExpectedTreesVisitor.java:
```diff
@@ -0,0 +1,421 @@
+    @Override
+    public Void visitLambdaExpression(LambdaExpressionTree tree, Void p) {
+        for (VariableTree parameter : tree.getParameters()) {
+            // Parameter types might not be specified for lambdas. When not specified, JavaParser
```
Make the actor clear in the first sentence:  it's the programmer.
Also, what does javac do in this case?  The second sentence mentions JavaParser.

mernst commented on framework/src/main/java/org/checkerframework/framework/ajava/ExpectedTreesVisitor.java:
```diff
@@ -0,0 +1,421 @@
+                hasVarType = true;
+            }
+
+            super.visit(node, p);
```
Once `hasVarType` is true, there is no need to do any more visiting.

(For efficiency, it would even be possible to shortcut, such as by throwing an exception that could be caught by the static `hasVarType` method that I added, and then you also wouldn't need the variable.  You could skip this optimization if you find it gross.)

jwaataja replied:
This class is only used by the testing framework, and this would only be an optimization if the file actually uses `var`, which doesn't occur often in the tests as far as I know.

mernst commented on framework/src/main/java/org/checkerframework/framework/ajava/JointJavacJavaParserVisitor.java:
```diff
@@ -0,0 +1,2382 @@
+public abstract class JointJavacJavaParserVisitor implements TreeVisitor<Void, Node> {
+    @Override
+    public Void visitAnnotation(AnnotationTree javacTree, Node javaParserNode) {
+        // javac stores annotation arguments as assignments, so @MyAnno("myArg") might be stored the
```
Should "might be" be "is"?  If not, under what circumstance is the statement about storage true?

mernst commented on framework/src/main/java/org/checkerframework/framework/ajava/JointJavacJavaParserVisitor.java:
```diff
@@ -0,0 +1,2382 @@
+        } else if (javaParserNode instanceof SingleMemberAnnotationExpr) {
+            SingleMemberAnnotationExpr node = (SingleMemberAnnotationExpr) javaParserNode;
+            processAnnotation(javacTree, node);
+            assert javacTree.getArguments().size() == 1;
```
I suggest that you also assert that the element name is "value".

mernst commented on framework/src/main/java/org/checkerframework/framework/ajava/JointJavacJavaParserVisitor.java:
```diff
@@ -0,0 +1,2382 @@
+            Iterable<? extends StatementTree> javacStatements,
+            Iterable<Statement> javaParserStatements) {
+        Iterator<? extends StatementTree> javacIter = javacStatements.iterator();
+        boolean hasNextJavac = javacIter.hasNext();
```
Here and elsewhere, you could drop the boolean and just test whether `javacStatement` is null.  That would be a bit shorter.  But you can leave this code as is if you prefer.

mernst commented on framework/src/main/java/org/checkerframework/framework/ajava/JointJavacJavaParserVisitor.java:
```diff
@@ -0,0 +1,2382 @@
+        if (javaParserNode instanceof ClassOrInterfaceDeclaration) {
+            ClassOrInterfaceDeclaration node = (ClassOrInterfaceDeclaration) javaParserNode;
+            processClass(javacTree, node);
+            if (javacTree.getTypeParameters().size() == node.getTypeParameters().size()) {
```
Should it be an error if these are not equal?

mernst commented on framework/src/main/java/org/checkerframework/framework/ajava/JointJavacJavaParserVisitor.java:
```diff
@@ -0,0 +1,2382 @@
+    public void visitAnonymouClassBody(
+            ClassTree javacBody, List<BodyDeclaration<?>> javaParserMembers) {
+        List<Tree> members = new ArrayList<>(javacBody.getMembers());
+        while (!members.isEmpty()) {
```
Why is this a `while` loop rather than an `if` test?  I think there can be at most one initializer to remove.

mernst commented on framework/src/main/java/org/checkerframework/framework/ajava/JointJavacJavaParserVisitor.java:
```diff
@@ -0,0 +1,2382 @@
+            javacTree.getExpression().accept(this, node.getExpression());
+        } else if (javaParserNode instanceof ExplicitConstructorInvocationStmt) {
+            // In this case the expression will be a MethodTree, which would be better to match with
+            // the statement than the expression statement itself.
```
I'm confused by this comment.  Should it be "the expression than the expression statement itself."?  In any event, please clarify.

mernst commented on framework/src/main/java/org/checkerframework/framework/ajava/JointJavacJavaParserVisitor.java:
```diff
@@ -0,0 +1,2382 @@
+        assert javacTree.getUpdate().size() == node.getUpdate().size();
+        Iterator<Expression> javaParserIter = node.getUpdate().iterator();
+        for (ExpressionStatementTree update : javacTree.getUpdate()) {
+            update.getExpression().accept(this, javaParserIter.next());
```
This seems to visit a javac expression and a JavaParser statement, which seems wrong to me because they are different kinds of AST nodes.  Perhaps clarifying the comment will resolve my confusion.

mernst commented on framework/src/main/java/org/checkerframework/framework/ajava/JointJavacJavaParserVisitor.java:
```diff
@@ -0,0 +1,2382 @@
+            throwUnexpectedNodeType(javacTree, javaParserNode, LambdaExpr.class);
+        }
+
+        LambdaExpr node = (LambdaExpr) javaParserNode;
```
Minor:  I would find it easier to read the code if variables like `node` were prefixed by `javaParser` or `javac`.  In this case, a clearer name would be `javaParserLambdaExpr`.  (Do you prefer the generic `node` so all method bodies use the same variable name for the downcasted variable?)

jwaataja replied:
Yes, I prefer that name because it can be used in every method, and the name `javaParserNode` is already used for the parameter type.

mernst commented on framework/src/main/java/org/checkerframework/framework/ajava/JointJavacJavaParserVisitor.java:
```diff
@@ -0,0 +1,2382 @@
+        // annotations and other modifiers, so the order doesn't match. It might be that for
+        // JavaParser, the annotations and other modifiers are also accessed separately.
+        javacTree.getReturnType().accept(this, javaParserNode.getType());
+        // Unlike other constructs, the list is non-null even if no type parameters are present.
```
Does "the list" refer to a javac or JavaParser data structure?

mernst commented on framework/src/main/java/org/checkerframework/framework/ajava/JointJavacJavaParserVisitor.java:
```diff
@@ -0,0 +1,2382 @@
+        if (javaParserNode instanceof MethodCallExpr) {
+            MethodCallExpr node = (MethodCallExpr) javaParserNode;
+            processMethodInvocation(javacTree, node);
+            // In javac, the type arguments will be empty even if no type arguments are specified,
```
Can you remove "even" here?

mernst reviewed (changes requested):
Overall this looks great.  Here are a few comments and questions.

mernst commented on framework/src/main/java/org/checkerframework/framework/ajava/JointJavacJavaParserVisitor.java:
```diff
@@ -0,0 +1,2386 @@
+        if (javaParserNode instanceof VariableDeclarator) {
+            VariableDeclarator node = (VariableDeclarator) javaParserNode;
+            processVariable(javacTree, node);
+            if (!node.getType().isVarType()
```
Could you write a comment about what is happening here?

mernst reviewed (changes requested):
Here are comments on `JointJavacJavaParserVisitor.java`.

jwaataja commented on framework/src/main/java/org/checkerframework/framework/ajava/JointJavacJavaParserVisitor.java:
```diff
@@ -0,0 +1,2386 @@
+                continue;
+            }
+
+            // In javac, a line like "int i = 0, j = 0" is expanded as two sibling VariableTree
```
The particular logic and methods called in each case are different, so abstracting out would probably complicate the code more than necessary. The code to advance the iterators was repetitive, so I've introduced a class `LookaheadIterator` to make it easier to read.

mernst replied:
Rather than defining your own `LookaheadIterator`, how about using Guava's `PeekingIterator` which I think provides the same functionality?

mernst reviewed (changes requested):
Oops, I hadn't posted these other comments.

mernst commented on framework/src/main/java/org/checkerframework/framework/ajava/ExpectedTreesVisitor.java:
```diff
@@ -0,0 +1,419 @@
+    @Override
+    public Void visitCompilationUnit(CompilationUnitTree tree, Void p) {
+        // When using the "var" keyword, javac replaces it with the correct type, meaning the
+        // inserted type has no corresponding javac tree. There is no way of knowing if the
```
Can you check whether it has a position?  For example:
```java
import com.sun.tools.javac.util.Position;
import com.sun.tools.javac.tree.JCTree.JCExpression;
public static boolean isVarDecl(VariableTree varTree) {
    return ((JCExpression)varTree.getType()).pos == Position.NOPOS;
}
```

mernst replied:
Do you mean "no corresponding JavaParser tree"?

mernst commented on framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java:
```diff
@@ -622,37 +696,55 @@ private void processPackage(PackageDeclaration packDecl) {
      * Process a type declaration: copy its annotations to {@code #annotationFileAnnos}.
      *
+     * <p>This method stores the declaration's type parameters in {@link #typeParameters}. When
+     * processing an ajava file, where traversal is handled externaly by a visitor, these type
```
Does "a visitor" specifically mean `AjavaParserVisitor`?  If possible, be specific.

mernst commented on framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java:
```diff
@@ -2432,6 +2536,66 @@ private void stubDebug(String warning) {
     }
 
+    /**
+     * Walks the javac tree for a file, processing the JavaParser node for each program element it
```
From the documentation and the name of this class, I thought it was an abstract base class intended to be subclassed.  Can you be specific about what processing it does, and give it a better name?  This seems to do special processing for constructs that can declare type variables (classes, methods), but I'm not sure why it overrides `visitVariable`.

mernst commented on framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java:
```diff
@@ -233,6 +243,16 @@
     /** Parses stub files and stores annotations from stub files. */
     public final AnnotationFileElementTypes stubTypes;
```
Should the documentation of these two fields mention that they only store public elements?

mernst commented on framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java:
```diff
@@ -233,6 +243,16 @@
+    public final AnnotationFileElementTypes ajavaTypes;
+
+    /**
+     * If processing a file, stores any annotations read from an ajava file for that class. Unlike
```
Should "processing a file" be "processing an ajava file"?  Or does this mean "if the Checker Framework is type-checking a .java file, stores annotations read from an ajava file for that class, if one exists"?

mernst commented on framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java:
```diff
@@ -480,13 +502,24 @@ public AnnotatedTypeFactory(BaseTypeChecker checker) {
                                     + " should be one of: -Ainfer=jaifs, -Ainfer=stubs");
             }
-            wholeProgramInference = new WholeProgramInferenceScenes(this);
+            if (checker.getOption("infer").equals("ajava")) {
```
Would it be cleaner to check the variable `wpiOutputFormat` here rather than calling `getOption` and `String.equals`?

mernst commented on framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java:
```diff
@@ -670,6 +703,58 @@ protected void initializeReflectionResolution() {
+     *
+     * @param root the new compilation unit to use
+     * @param shouldCheckVisitor if true, run the visitor that verifies that the javac tree can be
+     *     visited with its corresponding JavaParser AST. The check only occurs if
```
At first I thought this comment was explaining when `shouldCheckVisitor` is true.  But it is actually talking about a different boolean value.  What is the relationship between them, and why are both needed?  That is, why can't `shouldCheckeVisitor` be set according to `-AcheckJavaParserVisitor`?

mernst commented on framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java:
```diff
@@ -3562,6 +3703,34 @@ private AnnotationMirror getDeclAnnotation(
 
+        if (!ajavaTypes.isParsing()) {
+
+            // Retrieving annotations from ajava files.
```
This code block appears identically 3 times.  Please abstract it out into a method.

mernst reviewed (changes requested):
Here are more comments.

jwaataja commented on framework/src/main/java/org/checkerframework/common/wholeprograminference/WholeProgramInferenceJavaParserStorage.java:
```diff
@@ -0,0 +1,1076 @@
+        } else if (curATM.getKind() == TypeKind.TYPEVAR) {
+            // getExplicitAnnotations will be non-empty for type vars whose bounds are explicitly
+            // annotated.  So instead, only insert the annotation if there is not primary annotation
+            // of the same hierarchy.  #shouldIgnore prevent annotations that are subtypes of type
```
It's only relevant to the Scenes implementation, I've removed it here.

jwaataja commented on framework/src/main/java/org/checkerframework/common/wholeprograminference/WholeProgramInferenceJavaParserStorage.java:
```diff
@@ -0,0 +1,1076 @@
+    }
+
+    /**
+     * Reads in the source file containing {@code tree} and creates a wrapper around {@code tree}.
```
You're right in that it stores everything in `classToAnnos`.

> Please write the Javadoc in terms of the public abstractions.

This is a private method, so I think it's fine to mention the private fields here.

jwaataja commented on framework/src/main/java/org/checkerframework/common/wholeprograminference/WholeProgramInferenceJavaParserStorage.java:
```diff
@@ -0,0 +1,1076 @@
+                        if (!elt.getKind().isField()) {
+                            return;
+                        }
+
```
It does. There's a method `processClass` that takes an `EnumDeclaration` as its second argument. It's called `processClass` because javac stores the enum declaration as a `ClassTree`.

jwaataja replied:
We would only add enum constants if we did inference on them, which we're not doing for now.

jwaataja commented on framework/src/main/java/org/checkerframework/common/wholeprograminference/WholeProgramInferenceJavaParserStorage.java:
```diff
@@ -0,0 +1,1076 @@
+            return;
+        }
+
+        if (!methodDeclaration.getParentNode().isPresent()) {
```
I'm not aware of any cases where the parent node wouldn't be present. I'll remove the check and see if it still works.

jwaataja commented on framework/src/main/java/org/checkerframework/common/wholeprograminference/WholeProgramInferenceStorage.java:
```diff
@@ -0,0 +1,197 @@
+    public void setFileModified(String path);
+
+    /**
+     * Performs any preparation required for inference on Elements of a class. Should be called on
```
I think currently we only infer things for constructs with an `Element`, but that might change in the future so I'll remove the capitalization.

jwaataja commented:
I've addressed everything. This should be ready for another review.

smillst reviewed (commented):
Here are comments on the manual.  I'm reviewing the code now.

smillst reviewed (commented):
Here's my review of the changes to the framework classes.  Please also see, https://github.com/jwaataja/checker-framework/pull/11.  I'm going to skim the reset of the changes next.

jwaataja commented:
@smillst I've merged your pull request and addressed the comments so far. Please see the new paragraph at the bottom of the `ajava-using` section of the manual in `annotating-libraries.tex` to make sure it looks correct.

smillst commented on docs/manual/annotating-libraries.tex:
```diff
@@ -844,6 +844,69 @@
 in the version of the library you are using.
 
 
+\sectionAndLabel{Ajava Files}{ajava-files}
```
Should  https://checkerframework.org/manual/#whole-program-inference be updated to reference the new file format?

jwaataja replied:
Currently, the `wpi.sh` and `wpi-many.sh` scripts only support stub files. It should be easy to update them to use ajava files, but that can be done separately.

smillst reviewed (approved):
've skimmed the new files added and didn't see any needed changes.   The changes made in response to my other reviews look good, thanks! I

smillst commented:
@mernst should there be a change log entry for this?

mernst commented:
Thanks, Jason!
